### PR TITLE
feat: add more context to slack notifications from CDK pipelines

### PR DESCRIPTION
**Pull request by evan-boone**
This commit adds more context to Slack notifications from CDK pipelines, which are sent based upon status change events in CDK pipelines.

1. All Slack notifications from CDK pipelines include the execution ID from CodePipeline
2. If metadata is sent from CI in the trigger file, notifications will include information about which **branch** and **repo** CD was triggered from. They will also include the **commit hash** and **commit author**. 
3. Consuments can set the `notificationLevel` prop when configuring Slack notification through `addSlackNotification` to express which event types should be sent to Slack. **WARN** includes notifications related to the failure of a pipeline execution. **INFO** adds notifications for the success of a pipeline execution. **DEBUG** adds notifications for the start and superseding of a pipeline execution.

**Breaking changes for consuments:**
1. This PR removes the `alwaysShowSucceeded` prop that could previously be sent in to `addSlackNotification` to indicate that notifications should be sent for the success of a pipeline execution. Use `notificationLevel` instead.
2. This PR removes the `accountGroupName` and  `accountDescription` props that could previously be sent in to `addSlackNotification`. Use  accountFriendlyName instead to indicate a name for the pipeline.

### CODE REVIEW

**Review by stekern — COMMENTED** [review]
Nice. Overall this is functionally close to what we want 🚀

I have some specific in-line comments as well as some general comments below. Note that a lot of this has to do with the state of the Lambda code before you started extending it, but I think we should try to refactor and/or improve the code a bit to make it easier to test and maintain in the future, especially as the function is growing.
- The handler is doing a lot of different operations now, and the code can be somewhat hard to follow. Can we simplify this in some places by extracting specific logic to dedicated methods?
- I would try to prepare most Slack fields, at least the larger ones, before adding them to the final payload. This makes the code somewhat easier to reason about as we're not doing too many different operations at once. It also allows us to more easily extract the logic into dedicated methods that we can test.
- We're doing a lot of string formatting now, and this can be a bit brittle and cumbersome to maintain (e.g., duplicate whitespace, missing newlines, ...). Some of the string concatenation and removal of empty fields etc. can be simplified a bit by doing something like this

  ```python
  pretext = " ".join(s for s in [f"{emoji} Pipeline *<{pipeline_url}|{pipeline_name}>*",  f"*{state_text}*", account_friendly_name] if s)
  text = "\n".join(s for s in [f"*Execution:* <{execution_url}|{execution_id}>", text_for_failed] if s)
  ```
- We should generally try to clean up dead code (e.g., `account_text`).
- We should probably add some [type hints](https://docs.python.org/3/library/typing.html) and [docstrings](https://peps.python.org/pep-0257/#one-line-docstrings) where possible, at least for new methods.

File: assets/pipeline-slack-notification-lambda/index.py [hunk 1]
```diff
@@ -117,48 +156,65 @@ def handler(event, context):
...
 
     state_text = state
     if previous_failed and state == "SUCCEEDED":
         state_text += " (previously failed)"
 
-    blocks_for_failed = get_blocks_for_failed(pipeline_name, execution_id, state)
-
     account_text = account
     if ACCOUNT_DESC is not None:
         account_text += f" ({ACCOUNT_DESC})"
```
[the remark this answers is not in the document]
Comment (evan-boone) [reply assets/pipeline-slack-notification-lambda/index.py:177]: Seems like a good idea. I don't see the need for a description and group name, the friendly name should be enough.

File: assets/pipeline-slack-notification-lambda/index.py [hunk 2]
```diff
@@ -117,48 +156,65 @@ def handler(event, context):
...
+        if 'vcs' in ci_metadata:
+            repository_owner = ci_metadata['vcs']['repositoryOwner']
+            repository_name = ci_metadata['vcs']['repositoryName']
+            commit_hash = ci_metadata['vcs']['commitHash'][:7]
+            if repository_owner and repository_name and commit_hash:
+                github_commit_link = f"https://github.com/{repository_owner}/{repository_name}/commit/{commit_hash}"
+            if github_commit_link and ci_metadata['vcs']['branchName']:
+                commit_link_text = f"{repository_owner}/{repository_name}@{ci_metadata['vcs']['branchName']} ({commit_hash})"
+
+    emoji_prefix, message_color = get_message_formating_for_state(state)
```
[the remark this answers is not in the document]
Comment (evan-boone) [reply assets/pipeline-slack-notification-lambda/index.py:195]: Is this a better practice than having a method return the values?

File: assets/pipeline-slack-notification-lambda/index.py [hunk 3]
```diff
@@ -57,12 +76,16 @@ def get_previous_pipeline_execution(pipeline_name, execution_id):
...
 
 
-def get_blocks_for_failed(pipeline_name, execution_id, state):
-    # We only show details if the pipeline has completed with failed state.
-    # If we were to process this for other events such as started events,
-    # we would include details from after the event took place.
+def get_text_for_failed(pipeline_name: str, execution_id: str, state: str) -> str:
+    """Retrieve details from failed execution
+
+    We only show details if the pipeline has completed with failed state.
```
Comment (stekern) [inline assets/pipeline-slack-notification-lambda/index.py:82]: doc: Should this be a part of the method's documentation or be inline comments inside the method instead?

**Review by stekern — COMMENTED** [review]
Good stuff! I've added some various input. Let me know if you want to talk through any of the comments or if you want to do some pair programming to get this across the finish line 👨‍💻

File: src/cdk-pipelines/slack-notification.ts [hunk 4]
```diff
@@ -93,11 +97,20 @@ export class SlackNotification extends constructs.Construct {
       }),
     )
 
+    reportFunction.grantPrincipal.addToPrincipalPolicy(
+      new iam.PolicyStatement({
+        actions: ["s3:GetObject", "s3:ListBucket", "s3:GetObjectVersion"],
+        resources: [
+          `${props.artifactsBucket.bucketArn}/pipelines/${props.pipeline.pipelineName}/*`,
```
[the remark this answers is not in the document]
Comment (evan-boone) [reply src/cdk-pipelines/slack-notification.ts:104]: I had looked into it before, but I think I made some error. I got it to work now with _grantRead_

**Review by stekern — APPROVED** [review]
LGTM 🚀 Nice work.

Before merging, could we clean up the commit history a bit and update the PR title and description?
